### PR TITLE
New Component: Data Iterator (DataTable abstraction)

### PR DESCRIPTION
**Pull request by Phlow2001**
- Extracted data related behavior from VDataTable into data-iterable mixin
- Refactored VDataTable to be render and data table specific behavior with data-iterable mixin
- Created new VDataIterator component with data-iterable mixin
- Fixed bugs in VDataTable no-data and no-results, as well as fixing and enhancing data table tests.
- Added VDataIterator tests and snapshots

Example usage of VDataIterator: https://codepen.io/jayblanchard/pen/NvbwMa?editors=1010

### CODE REVIEW

File: src/components/VDataIterator/VDataIterator.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    },
+    genFooter () {
+      const children = []
+
+      if (this.$slots.footer) {
+        children.push(this.$slots.footer)
+      }
+
+      if (!this.hideActions) {
+        children.push(this.$createElement('div', this.genActions())
```
Comment (nekosaur) [inline src/components/VDataIterator/VDataIterator.js]: The div you create here should be unnecessary. `genActions` returns a root div.

File: src/components/VDataIterator/VDataIterator.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        'dataiterator': true,
+        'dataiterator--select-all': this.selectAll !== false,
+        'theme--dark': this.dark,
+        'theme--light': this.light
+      }
+    }
+  },
+
+  methods: {
+    createProps (item, index) {
```
Comment (nekosaur) [inline src/components/VDataIterator/VDataIterator.js]: Try to refactor createProps from v-data-table into data-iterable instead.

File: src/components/VDataIterator/VDataIterator.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    tag: {
+      type: String,
+      default: 'div'
+    }
+  },
+
+  computed: {
+    classes () {
+      return {
+        'dataiterator': true,
```
Comment (nekosaur) [inline src/components/VDataIterator/VDataIterator.js]: Use kebab-case for css class names

File: src/components/VDataIterator/VDataIterator.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+export default {
+  name: 'v-data-iterator',
+
+  mixins: [DataIterable],
+
+  inheritAttrs: false,
+
+  props: {
+    tag: {
```
Comment (nekosaur) [inline src/components/VDataIterator/VDataIterator.js]: Perhaps change this to `content-tag`, and add explicit `content-props` prop (Object) to be used with `v-bind=` on content element?

File: src/components/VDataIterator/VDataIterator.spec.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        items: [
+          { other: 1, col1: 'foo', col2: 'a', col3: 1 },
+          { other: 2, col1: null, col2: 'b', col3: 2 },
+          { other: 3, col1: undefined, col2: 'c', col3: 3 }
+        ]
+      }
+    }
+  }
+
+  it('should match a snapshot - no matching records', () => {
```
Comment (nekosaur) [inline src/components/VDataIterator/VDataIterator.spec.js]: It's probably a good idea to check in these tests that the correct text is actually displayed.

**Comment by nekosaur** [comment]
Try to eliminate the last identical code block that codebeat complains about

File: src/components/VDataIterator/VDataIterator.js [hunk 6]
```diff
@@ -49,38 +49,24 @@ export default {
...
+      }, content)]
     },
     genFilteredItems () {
+      if (!this.$scopedSlots.item) {
+        return null
+      }
+
       const items = []
-      this.filteredItems.forEach((item, index) => {
+      for (let index = 0, len = this.filteredItems.length; index < len; ++index) {
```
Comment (Phlow2001) [inline src/components/VDataIterator/VDataIterator.js]: https://jsperf.com/native-map-versus-array-looping

Native loop outperforms by a good bit in most cases, and by a lot in some cases (like latest safari on ios)

**Pull request merged**